python/lsst/atmospec/nightAnalysis.py

```diff
@@ -78,56 +73,61 @@ def getLineValue(table, lineName, columnName, nanForMissingValues=True):
     return table[columnName][rowNum]
 
 
-class NightStellarSpectra():
+class NightStellarSpectra:
     """Class for holding the results for a night's observations of a given star
     """
 
-    def __init__(self, rerunDir, dayObs, targetName, *, butler=None, ignoreSeqNums=[]):
+    def __init__(self, rerunDir, dayObs, targetName, instrument, *, butler=None, ignoreSeqNums=[],
+                 collections=None):
         self.rerunDir = rerunDir
-        self.dayObs = dayObs
+        self.dayObs = int(dayObs)
         self.targetName = targetName
+        if instrument is None:
+            raise ValueError("An instrument name must be provided.")
+        self.instrument = instrument
+
         if butler:
-            self.butler = butler
+            self.butler = dafButler.Butler(butler=butler, instrument=instrument, collections=collections)
         else:
-            self.butler = dafPersist.Butler(rerunDir)
+            self.butler = dafButler.Butler(rerunDir, instrument=instrument, collections=collections)
         self._loadExtractions(ignoreSeqNums)
         # xxx maybe just load everything and then call removeSeqNums()?
 
     def isDispersed(self, seqNum):  # TODO: change this to use the isDispersedDataId in utils
         """Match object and check is dispersed"""
-        filt = self.butler.queryMetadata('raw', 'filter', dayObs=self.dayObs, seqNum=seqNum)[0]
+        where = "exposure.day_obs = dayObs and exposure.seq_num = seqNum"
+        records = list(self.butler.registry.queryDimensionRecords("exposure",
+                                                                  where=where,
+                                                                  bind={"dayObs": self.dayObs,
+                                                                        "seqNum": seqNum},
+                                                                  instrument=self.instrument,
+                                                                  ))
+        filt = records[0].physical_filter
```

Comment:
I think it would be neater here to do as the TODO says, and make a dataId with `self.dayObs` and the `seqNum` passed in and just pass that to `isDispersedDataId()` in utils. The code in there is functionally identical, so I think it's best to have this sort of stuff live in one place when possible.

Reply:
I hadn't looked at any of the utility code so was replacing what was there without any idea about alternatives.

Reply:
The two functions don't have the same logic. The bit to get the record is the same but one checks for `ronchi` in the name and the other checks for grating not saying EMPTY.

Reply:
I have moved it over and you can resolve the subtle difference between `is ronchi` and `is not empty`.

Reply:
Just confirming for posterity here that the utils way is the right way, further motivating putting it all in one place (this was from before we had a hologram).